jupyterhub/templates/_helpers-names.tpl

```diff
@@ -74,20 +74,23 @@
     {{- include "jupyterhub.fullname.dash" . }}hub
 {{- end }}
 
-{{- /* hub Secret */}}
-{{- define "jupyterhub.hub-secret.fullname" -}}
+{{- /* hub-existing-secret Secret */}}
+{{- define "jupyterhub.hub-existing-secret.fullname" -}}
     {{- /* A hack to avoid issues from invoking this from a parent Helm chart. */}}
     {{- $existing_secret := .Values.hub.existingSecret }}
     {{- if ne .Chart.Name "jupyterhub" }}
         {{- $existing_secret = .Values.jupyterhub.hub.existingSecret }}
```

Comment:
Should the chart name change where in values our secret is? I wouldn't think a nested .Values.jupyterhub should be allowed to exist in any event.

Reply:
`.Chart.name` is based on the name in Chart.yaml which will always be `"jupyterhub"` unless this template was called from a parent helm chart. So this logic is an attempt to make this return a valid response even if the passed scope must now dig into `.Values.jupyterhub` instead of just `.Values` to find what's relevant.

The point of doing this is to help a Helm chart like DaskHub or similar not hardcode names of resources it wants to interact with in this chart.

Reply:
But that's not how parent charts work, is it? The the parent passes .Values.$dependency-name *as* .Values to the dependency chart. A chart doesn't need to know its name to find its values, otherwise ~nothing in this chart would work in binderhub.

Reply:
All the charts templates will work fine and won't see `.Values.jupyterhub.hub` etc but instead `.Values.hub`. But, templates are globally scoped and can be used from a parent chart with the parent chart's scope.

So, in a daskhub template, you could do `{{ include "jupyterhub.proxy-public.fullname" . }}` thanks to this, because when that template gets the scope's `.Chart.Name` is `"daskhub"` and now have `.Values.jupyterhub.hub`, the globally accessible named template from the jupyterhub helm chart will render as if it was called with a JupyterHub scope even though it wasn't.

Reply:
> But, named templates defined in _helpers.tpl are globally scoped and can be used from a parent chart with the parent chart's scope.

Ah! This is the part I misunderstood, thank you! It's a specific issue to *exporting* helpers to be used by downstream charts. I never realized helpers were available to other charts. Do we need to think about what our public helper API is and how it's documented? I always assumed these were purely private. What would happen in  a case like mybinder.org, where jupyterhub is a dependency two hops away, i.e. `binderhub.jupyterhub.hub...`? Is that sustainable?

Reply:
I responded in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/2057!